DA/Internal/Desugar.hs

```diff
@@ -109,6 +109,27 @@ class HasToAnyChoice t c r | t c -> r where
 class HasFromAnyChoice t c r | t c -> r where
   _fromAnyChoice : proxy t -> Any -> Optional c
 
+class HasKey t k | t -> k where
+  key : t -> k
+
+class HasLookupByKey t k | t -> k where
+  lookupByKey : k -> Update (Optional (ContractId t))
+
+class HasFetchByKey t k | t -> k where
+  fetchByKey : k -> Update (ContractId t, t)
+
+class HasMaintainer t k | t -> k where
+  _maintainer : proxy t -> k -> [Party]
+
+class HasToAnyContractKey t k | t -> k where
+  _toAnyContractKey : proxy t -> k -> Any
+
+class HasFromAnyContractKey t k | t -> k where
+  _fromAnyContractKey : proxy t -> Any -> Optional k
+
+class HasExerciseByKey t k c r | t -> k, t c -> r where
+  exerciseByKey : proxy t -> c -> Update r
```

Comment:
This seems off. Shouldn’t you pass in the key here?

Reply:
ah good point! For reference, the current signature is

```haskell
exerciseByKey : (HasFetchByKey t k, HasExercise t c r) => k -> c -> Update r
```

I'm not sure if we need/want the `proxy t` argument, it seems the current one works fine with the `t` as a type (rather than term) argument - `exerciseByKey @MyTemplate myKey myChoice`

Reply:
I think you might need to be careful that you don’t have to turn on `AllowAmbiguousTypes`.